grails-app/services/org/pih/warehouse/product/ProductMergeService.groovy

```diff
@@ -0,0 +1,375 @@
+
+            // Swap product
+            obsoleteReceiptItem.product = primary
+
+            // Swap inventory item (in case the primary product had the same lot as obsolete)
+            def obsoleteLotNumber = obsoleteReceiptItem.inventoryItem?.lotNumber
+            def primaryInventoryItem = primary.inventoryItems?.find {
+                it.lotNumber == obsoleteLotNumber
+            }
+            obsoleteReceiptItem.inventoryItem = primaryInventoryItem
+            // Note: needs flush because of "User.locationRoles not processed by flush"
+            obsoleteReceiptItem.save(flush: true)
+        }
+
+        // 2.5 TransactionEntry
+        List<TransactionEntry> obsoleteTransactionEntries = TransactionEntry.findAllByProduct(obsolete)
+        obsoleteTransactionEntries?.each { TransactionEntry obsoleteTransactionEntry ->
+            logProductMergeData(primary, obsolete, obsoleteTransactionEntry)
+
+            // Swap product
+            obsoleteTransactionEntry.product = primary
+
+            // Swap inventory item (in case the primary product had the same lot as obsolete)
+            def obsoleteLotNumber = obsoleteTransactionEntry.inventoryItem?.lotNumber
+            def primaryInventoryItem = primary.inventoryItems?.find {
+                it.lotNumber == obsoleteLotNumber
+            }
+            obsoleteTransactionEntry.inventoryItem = primaryInventoryItem
+            // Note: needs flush because of "User.locationRoles not processed by flush"
+            obsoleteTransactionEntry.save(flush: true)
+        }
+
+        /**
+         * FINAL PROCESSING
+         * */
+
+        // 1. Deactivate obsoleted product
+        obsolete.active = false
+        if (!obsolete.save(flush: true)) {
+            throw new Exception("Cannot save obsolete product due to: " + obsolete.errors?.toString())
+        }
+        if (!primary.save(flush: true)) {
+            throw new Exception("Cannot save primary product due to: " + primary.errors?.toString())
+        }
+
+        // 2. Trigger refresh of product_demand, stockout_fact, product_summary views
+        // Skipped for now because demand data might take too long and other are not that relevant (plus InventorySnapshot
+        // and ProductAvailability are handled during inventory items swap).
+        // TODO / FIXME: Consider running these sequentionally in a separate job if needed
+        // RefreshStockoutDataJob.triggerNow()
+        // RefreshDemandDataJob.triggerNow() // <- FIXME: probably can be skipped if takes too long time to process
+        // RefreshOrderSummaryJob.triggerNow() // <- FIXME: maybe do a refresh by list of affected orders instead of full table
+    }
+
+    void logProductMergeData(Product primary, Product obsolete, def relatedObject) {
+        ProductMergeLogger productMergeLogger = new ProductMergeLogger(
+            primaryProduct: primary,
+            obsoleteProduct: obsolete,
+            relatedObjectId: relatedObject?.id,
```

Comment:
@jmiranda Looking into the number of records produced, I think this would be better to be a `LONGTEXT` field for `relatedObjectIds` with a common delimiter like a comma or sth.

Reply:
Ugh. It's feeling a lot like we need a better / different storage mechanism for this i.e. document store.